spec/packer_spec.rb

```diff
@@ -589,4 +589,31 @@ def to_msgpack_ext
       GC.stress = stress
     end
   end
+
+  it "doesn't crash when using recursive packers concurrently" do
+    hash_with_indifferent_access = Class.new(Hash)
+    msgpack = MessagePack::Factory.new
+    msgpack.register_type(
+      0x02,
+      hash_with_indifferent_access,
+      packer: ->(value, packer) do
+        packer.write("a".b * 600_0000) # Over MSGPACK_BUFFER_STRING_WRITE_REFERENCE_DEFAULT
```

Comment:
So this is what was likely missing to repro, on writes msgpack only reference the original string if it's over 500kiB, which is huge and rarely happens.

We should consider reducing that default very significantly in the future.